# **CARGAR DATOS E IMPORTAR BIBLIOTECAS NECESARIAS**

In [1]:
# BIBLIOTECAS NECESARIAS

# Bibliotecas para operaciones numéricas y manejo de datos
import numpy as np  # Operaciones numéricas
import pandas as pd  # Manejo de datos en formato tabular

# División del conjunto de datos y validación cruzada
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV

# Modelos de clasificación
from sklearn.ensemble import RandomForestClassifier

# Importar todo el módulo de métricas
import sklearn.metrics

# Métricas de evaluación y reportes
from sklearn.metrics import (
    confusion_matrix, ConfusionMatrixDisplay,
    PrecisionRecallDisplay,
    roc_curve, auc,
    classification_report,
    accuracy_score, precision_score, recall_score, f1_score
)

# Bibliotecas para gráficos
import matplotlib.pyplot as plt  # Gráficos básicos
import seaborn as sns  # Gráficos estadísticos, con menos código que matplotlib
import plotly.express as px  # Gráficos interactivos con interfaz sencilla y de alto nivel
import plotly.graph_objects as go  # Gráficos interactivos con interfaz más detallada y de bajo nivel

# Herramientas para iteradores
import itertools  # Herramientas funcionales para crear y utilizar iteradores

# Herramientas para preprocesamiento de datos y evaluación de modelos
from sklearn.preprocessing import StandardScaler  # Escalado de características para normalizar datos
from sklearn.model_selection import learning_curve  # Generación de curvas de aprendizaje para evaluar el rendimiento del modelo
from sklearn.model_selection import validation_curve  # Generación de curvas de validación para evaluar el rendimiento del modelo

In [2]:
# Cargar el dataset desde un archivo CSV
trans_fraud = pd.read_csv('C:\\Users\\mmene\\OneDrive\\Escritorio\\RESULTADOS PROYECTO\\DATA\\credit_card_fraud_detection.csv')

Se utilizará el data set de ***Credit Card Fraud Detection*** de ***Kaggle***: https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud

Este conjunto de datos evidencia transacciones anónimas con tarjeta de crédito etiquetadas como ***fraudulentas*** o ***auténticas***

# **DESCRIPCIÓN DE CARACTERÍSTICAS**

*   [ V1, V2,... V28 ] : Son los componentes principales resultado de una *transformación PCA*. Además, no es posible obtener más información de fondo sobre los datos, ni de estas características.

*   ' Time ' : Segundos transcurridos entre cada transacción y la primera transacción del conjunto de datos.

*   ' Amount ' : Cantidad de dinero retirado en cada transacción.

*   ' Class ' : Variable de respuesta. Toma el valor de 1 en caso de fraude y 0 en caso contrario.






In [3]:
trans_fraud.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   
5  0.260314 -0.568671  ... -0.208254 -0.559825 -0.026398 -0.371427 -0.232794   
6  0.081213  0.464960  ... -0.167716 -0.270710 -0.154104 -0.780055  0.750137   
7 -3.807864  0.615375  ...  1.943465 -1.015455  0.057504 -0.649709 -0.415267   
8  0.851084 -0.392048  ... -0.073425 -0.268092 -0.204233  1.011592  0.373205   
9  0.069539 -0.736727  ... -0.246914 -0.633753 -0.120794 -0.385050 -0.069733   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  
5  0.105915  0.253844  0.081080    3.67      0  
6 -0.257237  0.034507  0.005168    4.99      0  
7 -0.051634 -1.206921 -1.085339   40.80      0  
8 -0.384157  0.011747  0.142404   93.20      0  
9  0.094199  0.246219  0.083076    3.68      0  

[10 rows x 31 columns]

In [4]:
# Paso 1: Separar características (X) y la columna objetivo (y)
X = trans_fraud.drop(['Class'], axis=1)  # Eliminar la columna 'Class'
y = trans_fraud['Class']  # Conservar la columna 'Class'

# Paso 2: Crear un objeto StandardScaler
scaler = StandardScaler()

# Paso 3: Ajustar el escalador a las características y transformarlas
scaled = scaler.fit_transform(X)

# Paso 4: Convertir X_scaled de vuelta a un DataFrame
X_scaled = pd.DataFrame(scaled, columns=X.columns)

# Paso 5: Reunir X_scaled con la columna 'Class'
trans_fraud_scaled = pd.concat([X_scaled, y.reset_index(drop=True)], axis=1)

# Paso 6: Mostrar las primeras filas del DataFrame transformado
print(trans_fraud_scaled.head())

       Time        V1        V2        V3        V4        V5        V6  \
0 -1.996583 -0.694242 -0.044075  1.672773  0.973366 -0.245117  0.347068   
1 -1.996583  0.608496  0.161176  0.109797  0.316523  0.043483 -0.061820   
2 -1.996562 -0.693500 -0.811578  1.169468  0.268231 -0.364572  1.351454   
3 -1.996562 -0.493325 -0.112169  1.182516 -0.609727 -0.007469  0.936150   
4 -1.996541 -0.591330  0.531541  1.021412  0.284655 -0.295015  0.071999   

         V7        V8        V9  ...       V21       V22       V23       V24  \
0  0.193679  0.082637  0.331128  ... -0.024923  0.382854 -0.176911  0.110507   
1 -0.063700  0.071253 -0.232494  ... -0.307377 -0.880077  0.162201 -0.561131   
2  0.639776  0.207373 -1.378675  ...  0.337632  1.063358  1.456320 -1.138092   
3  0.192071  0.316018 -1.262503  ... -0.147443  0.007267 -0.304777 -1.941027   
4  0.479302 -0.226510  0.744326  ... -0.012839  1.100011 -0.220123  0.233250   

        V25       V26       V27       V28    Amount  Class  
0  0.24

In [6]:
# Guardar en una ubicación específica
trans_fraud_scaled.to_csv('C:\\Users\\mmene\\OneDrive\\Escritorio\\TESIS\\trans_fraud_scaled.csv', index=False)

# **ANÁLISIS EXPLORATORIO**

In [6]:
# Dimensiones del DataFrame (#FILAS,#COLUMNAS)
trans_fraud.shape

(284807, 30)

In [7]:
# Nombres de las columnas del DataFrame. Al agregar .values,
# se convierte este objeto Index en un array de Numpy que contiene solo los nombres de las columnas.
trans_fraud.columns.values

array(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'], dtype=object)

In [8]:
# Información acerca de las características del DataFrame
trans_fraud.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 30 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   V1      284807 non-null  float64
 1   V2      284807 non-null  float64
 2   V3      284807 non-null  float64
 3   V4      284807 non-null  float64
 4   V5      284807 non-null  float64
 5   V6      284807 non-null  float64
 6   V7      284807 non-null  float64
 7   V8      284807 non-null  float64
 8   V9      284807 non-null  float64
 9   V10     284807 non-null  float64
 10  V11     284807 non-null  float64
 11  V12     284807 non-null  float64
 12  V13     284807 non-null  float64
 13  V14     284807 non-null  float64
 14  V15     284807 non-null  float64
 15  V16     284807 non-null  float64
 16  V17     284807 non-null  float64
 17  V18     284807 non-null  float64
 18  V19     284807 non-null  float64
 19  V20     284807 non-null  float64
 20  V21     284807 non-null  float64
 21  V22     28

**ANÁLISIS 1:**

Se corrobora que el dataframe contiene **284807 transacciones**, de las cuales cada una posee **30 características** y **1 variable respuesta**.
No se presenta valores nulos, los valores de las características se presentan en decimal y la clase perteneciente es entera.

# **COMPRENDER VARIABLE OBJETIVO**

In [9]:
# VALORES ÚNICOS DE LA COLUMNA 'Class'
# 1. Crea un conjunto de los valores únicos presentes en la columna 'Class' del DataFrame
# 2. Muestra los valores únicos

value_etq = set(trans_fraud['Class'])
print("\033[1mVARIABLE RESPUESTA:\033[0m")
print("\nToma el valor de \033[1m1\033[0m en caso de \033[1mfraude\033[0m (transacciones fraudulentas) y toma el valor de \033[1m0\033[0m (transacciones auténticas)\nen \033[1mcaso contrario\033[0m\n")
print(f'Clases: {value_etq}')

VARIABLE RESPUESTA:

Toma el valor de 1 en caso de fraude (transacciones fraudulentas) y toma el valor de 0 (transacciones auténticas)
en caso contrario

Clases: {0, 1}


In [10]:
# Explorar la variable objetivo para entender la distribución de los valores dentro de una columna categórica.
# ".value_counts()" es un método que se utiliza para contar la cantidad de ocurrencias únicas de cada valor en una serie
trans_fraud['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

**ANÁLISIS 2:**

En esta base de datos solo existen **2 clases**. La cual la predicción toma el valor de **1** en caso de fraude (**transacciones fraudulentas**) y toma el valor de **0** (**transacciones auténticas**)
en caso contrario.
Adicionalmente, las clases presentan ***desbalance*** debido a que la clase mayoritaria contiene la mayoría de observaciones (284315), mientras que la clase de interés, que es la 1: transacciones fraudulentas, solo posee 492 observaciones.

# **VISUALIZACIÓN**: Gráfico de PASTEL

In [79]:
# Calcula el recuento de las clases de transacción (no fraudulenta y fraudulenta) y los ordena por índice
class_counts = trans_fraud['Class'].value_counts().sort_index()

# Extrae los valores (recuentos) de las clases
valores = class_counts.values

# Define etiquetas para las clases (0: no fraudulenta, 1: fraudulenta)
# Estas etiquetas se utilizarán en el gráfico para representar cada segmento
etiq =  {0: f"Transacción no fraudulenta", 1: f"Transacción fraudulenta"}

# Crea un gráfico de pastel con Plotly Express
# 'values' representa el tamaño de cada segmento, y 'names' proporciona las etiquetas para cada segmento
fig = px.pie(trans_fraud, values=valores, names=etiq,
             title='<b>DISTRIBUCIÓN DE CLASES POR TIPO DE TRANSACCIÓN</b>',
             color_discrete_sequence=['#522258', '#FFE31A'],
             )

# Ajusta el tamaño del texto del porcentaje en el gráfico para mejorar la visibilidad
fig.update_traces(textinfo='percent', textfont_size=15)

# Centra el título y coloca la leyenda debajo del gráfico
fig.update_layout(
    title_x=0.5,  # Centra el título
    title_y=0.90,  # Ajusta la posición vertical del título
    showlegend=True,  # Activa la leyenda
    legend_orientation="h",  # Pone la leyenda horizontal
    legend_y=-0.05,  # Coloca la leyenda debajo del gráfico
    legend_x=0.2
)

# Muestra el gráfico
fig.show()

**ANÁLISIS 3:**

Se observa que el **dataset se encuentra altamente desbalanceado**. Evidenciando que la clase minoritaria de interés solo representa **0,173%** del total de los datos. Esto se considera un caso de desequilibrio extremo, lo presenta (Malek,2023) en la clasificación dada según la gravedad del desbalance:

- Proporción de la clase minoritaria para un ***desequilibrio leve*** = [20%-40%]
- Proporción de la clase minoritaria para un ***desequilibrio moderado*** = [<20%]
- Proporción de la clase minoritaria para un ***desequilibrio extremo*** = [<1%]

# MODELO ENSAMBLADO: **RANDOM FOREST**

In [16]:
# Separar las características (features) y la variable objetivo (target) del DataFrame
X = trans_fraud.drop(['Class'], axis=1)  # Eliminar la columna 'Class' para obtener las características
y = trans_fraud['Class']  # Seleccionar la columna 'Class' como la variable objetivo

In [6]:
# 'HOLDOUT METHOD' para evaluar el rendimiento de un modelo de aprendizaje automático
# Este método divide el conjunto de datos en entrenamiento y prueba con diferentes tamaños de entrenamiento
# para evaluar cómo el rendimiento del modelo varía con el tamaño del conjunto de entrenamiento.

# Definir los diferentes tamaños para el conjunto de entrenamiento
train_sizes = [0.75,0.80,0.90]

# Crear una lista para almacenar los resultados de los diferentes conjuntos de entrenamiento y prueba
split_results = []

# Para cada tamaño de entrenamiento en train_sizes
for splits in train_sizes:
  # Se divide los datos X e y en conjuntos de entrenamiento y prueba con el tamaño especificado
  X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=splits, random_state=42)
  split_results.append((X_train, X_test, y_train, y_test))

# Mostrar las dimensiones de los conjuntos de entrenamiento y prueba para cada tamaño de entrenamiento
for i, (X_train, X_test, y_train, y_test) in enumerate(split_results):
  print(f"\033[1mConjunto con train_size={train_sizes[i]}\033[0m")
  print(f"Dimensión total es: {trans_fraud.shape}")
  print(f"X_train dimensión: {X_train.shape}, X_test dimensión: {X_test.shape}")
  print(f"y_train dimensión: {y_train.shape}, y_test dimensión: {y_test.shape}")
  print()

Conjunto con train_size=0.75
Dimensión total es: (284807, 31)
X_train dimensión: (213605, 30), X_test dimensión: (71202, 30)
y_train dimensión: (213605,), y_test dimensión: (71202,)

Conjunto con train_size=0.8
Dimensión total es: (284807, 31)
X_train dimensión: (227845, 30), X_test dimensión: (56962, 30)
y_train dimensión: (227845,), y_test dimensión: (56962,)

Conjunto con train_size=0.9
Dimensión total es: (284807, 31)
X_train dimensión: (256326, 30), X_test dimensión: (28481, 30)
y_train dimensión: (256326,), y_test dimensión: (28481,)



In [14]:
# MODELO RANDOM FOREST

# Lista para almacenar el resultado del modelo para cada conjunto de entrenamiento/prueba
models_split_results = []

# Iterar sobre los conjuntos generados
for i, (X_train, X_test, y_train, y_test) in enumerate(split_results):

  # Definir las métricas de evaluación
  scores = ["precision", "recall", "roc_auc", "f1"]

  # Mostrar información sobre el conjunto actual
  print(f"\033[1mConjunto con train_size={train_sizes[i]}:\033[0m")

  # Crear el modelo RANDOM FOREST
  modeloRF = RandomForestClassifier(
      n_jobs = -1, # Utilizar todos los núcleos disponibles para acelerar el entrenamiento
      random_state = 42 # Semilla para la reproducibilidad de los resultados
      )

  # Definir los parámetros para GridSearchCV
  grid_search = {
      "n_estimators": [10,12,14,16,18,20,50,100,200], # Número de árboles en el bosque aleatorio
      "criterion": ["gini", "entropy"], # Criterio para la selección del umbral de
      "max_features": ["sqrt","log2"], # Número máximo de características a considerar
      "max_depth": [10,20,30], # Profundidad máxima de cada
      "min_samples_split": [2,5,10], # Número mínimo de muestras requeridas para dividir un nodo interno
      "min_samples_leaf": [1,2,4] # Número mínimo de muestras requeridas en cada hoja
      }

  # Medir tiempo de ejecución
  start = pd.Timestamp.now()

  # Configuración de GridSearchCV
  grid = GridSearchCV(
      modeloRF,
      grid_search,
      cv=3, # Número de particiones en la validación cruzada
      refit=False, # No volver a entrenar el mejor modelo encontrado en toda la búsqueda
      scoring=scores, # Métricas de evaluación a usar
      verbose = 3, # Mostrar información adicional durante la búsqueda
      return_train_score=True # Devolver los resultados de entrenamiento
      )

  # Ejecutar GridSearchCV para encontrar el mejor modelo
  model_grid = grid.fit(X_train, y_train)

  # Medir tiempo de ejecución de GridSearchCV
  end = pd.Timestamp.now()
  print(f"Duración búsqueda: {end-start}")

  # Almacenar los resultados en la list 'models_split_results'
  models_split_results.append(model_grid.cv_results_)

  # Guardar modelos en un DataFrame
modelos_GSCV1= pd.DataFrame(models_split_results)

model_75 = models_split_results[0]
model_80 = models_split_results[1]
model_90 = models_split_results[2]

#75
# Crear una lista de diccionarios, donde cada diccionario corresponde a una ejecución
ajuste_modelo_75 = []

# Recorre cada índice 'i' basado en la longitud de los tiempos medios de ajuste ('mean_fit_time')
for i in range(len(model_75['mean_fit_time'])):
  # Crea un diccionario 'execution' para la ejecución actual, extrayendo el valor correspondiente al índice 'i' de cada clave en 'resultados_modelo_75'
  execution = {key: value[i] for key, value in model_75.items()}
  # Añade el diccionario de resultados de la ejecución actual a la lista 'ajuste_modelo_80'
  ajuste_modelo_75.append(execution)

# Convierte la lista de diccionarios 'ajuste_modelo_75' en un DataFrame de pandas para facilitar su manipulación y análisis
modelos_75 = pd.DataFrame(ajuste_modelo_75)
# Guarda el DataFrame resultante en un archivo Excel, permitiendo almacenar los resultados de manera estructurada
modelos_75.to_excel('C:\\Users\\mmene\\OneDrive\\Escritorio\\RESULTADOS PROYECTO\\MODELOS\\Random Forest\\credit_card_fraud_detection (BD1)\\modelos_75_ONLY_NORMALIZADO.xlsx')


#80
# Crear una lista de diccionarios, donde cada diccionario corresponde a una ejecución
ajuste_modelo_80 = []

# Recorre cada índice 'i' basado en la longitud de los tiempos medios de ajuste ('mean_fit_time')
for i in range(len(model_80['mean_fit_time'])):
  # Crea un diccionario 'execution' para la ejecución actual, extrayendo el valor correspondiente al índice 'i' de cada clave en 'resultados_modelo_75'
  execution = {key: value[i] for key, value in model_80.items()}
  # Añade el diccionario de resultados de la ejecución actual a la lista 'ajuste_modelo_80'
  ajuste_modelo_80.append(execution)

# Convierte la lista de diccionarios 'ajuste_modelo_80' en un DataFrame de pandas para facilitar su manipulación y análisis
modelos_80 = pd.DataFrame(ajuste_modelo_80)
# Guarda el DataFrame resultante en un archivo Excel, permitiendo almacenar los resultados de manera estructurada
modelos_80.to_excel('C:\\Users\\mmene\\OneDrive\\Escritorio\\RESULTADOS PROYECTO\\MODELOS\\Random Forest\\credit_card_fraud_detection (BD1)\\modelos_80_ONLY_NORMALIZADO.xlsx')


#90
# Crear una lista de diccionarios, donde cada diccionario corresponde a una ejecución
ajuste_modelo_90 = []

# Recorre cada índice 'i' basado en la longitud de los tiempos medios de ajuste ('mean_fit_time')
for i in range(len(model_90['mean_fit_time'])):
  # Crea un diccionario 'execution' para la ejecución actual, extrayendo el valor correspondiente al índice 'i' de cada clave en 'resultados_modelo_75'
  execution = {key: value[i] for key, value in model_90.items()}
  # Añade el diccionario de resultados de la ejecución actual a la lista 'ajuste_modelo_80'
  ajuste_modelo_90.append(execution)

# Convierte la lista de diccionarios 'ajuste_modelo_90' en un DataFrame de pandas para facilitar su manipulación y análisis
modelos_90 = pd.DataFrame(ajuste_modelo_90)
# Guarda el DataFrame resultante en un archivo Excel, permitiendo almacenar los resultados de manera estructurada
modelos_90.to_excel('C:\\Users\\mmene\\OneDrive\\Escritorio\\RESULTADOS PROYECTO\\MODELOS\\Random Forest\\credit_card_fraud_detection (BD1)\\modelos_90_ONLY_NORMALIZADO.xlsx')

Conjunto con train_size=0.75:
Fitting 3 folds for each of 972 candidates, totalling 2916 fits
[CV 1/3] END criterion=gini, max_depth=10, max_features=sqrt, min_samples_leaf=1, min_samples_split=2, n_estimators=10; f1: (train=0.908, test=0.838) precision: (train=0.995, test=0.932) recall: (train=0.834, test=0.762) roc_auc: (train=0.992, test=0.939) total time=   1.6s
[CV 2/3] END criterion=gini, max_depth=10, max_features=sqrt, min_samples_leaf=1, min_samples_split=2, n_estimators=10; f1: (train=0.905, test=0.820) precision: (train=0.995, test=0.958) recall: (train=0.829, test=0.717) roc_auc: (train=0.994, test=0.968) total time=   1.6s
[CV 3/3] END criterion=gini, max_depth=10, max_features=sqrt, min_samples_leaf=1, min_samples_split=2, n_estimators=10; f1: (train=0.905, test=0.842) precision: (train=0.995, test=0.886) recall: (train=0.830, test=0.802) roc_auc: (train=0.982, test=0.971) total time=   1.5s
[CV 1/3] END criterion=gini, max_depth=10, max_features=sqrt, min_samples_leaf=1,